pyanaconda/mutter_display.py

```diff
@@ -0,0 +1,165 @@
+
+
+class Monitor(object):
+    """Represent a connected physical monitor."""
+
+    def __init__(self, props):
+        self.id = MonitorId(props[0])
+        self.modes = list(map(lambda p: MonitorMode(p), props[1]))
+        self.properties = props[2]
+
+
+class LogicalMonitor(object):
+    """Represent the current logical monitor configuration"""
+
+    def __init__(self, props):
+        self.x = props[0]
+        self.y = props[1]
+        self.scale = props[2]
+        self.transform = props[3]
+        self.primary = props[4]
+        self.monitor_ids = list(map(lambda p: MonitorId(p), props[5]))
+        self.properties = props[6]
+
+
+class LogicalMonitorConfig(object):
+    """Logical monitor configuration object"""
+
+    def __init__(self, logical_monitor, monitors, x, y, width, height):
+        """Creates a LogicalMonitorConfig setting the given resolution if available."""
+        self._logical_monitor = logical_monitor
+        self._monitors = monitors
+
+        self.x = x
+        self.y = y
+        self.scale = logical_monitor.scale
+        self.transform = logical_monitor.transform
+        self.primary = logical_monitor.primary
+
+        self.monitors = list()
+        for monitor_id in logical_monitor.monitor_ids:
+            connector = monitor_id.connector
+            mode_id = self._get_matching_monitor_mode_id(monitors, monitor_id, width, height)
+            self.monitors.append((connector, mode_id, {}))
+
+    def _get_matching_monitor_mode_id(self, monitors, monitor_id, width, height):
+        monitor = next(filter(lambda m: m.id == monitor_id, monitors))
+        for mode in monitor.modes:
+            if mode.width == width and mode.height == height:
+                return mode.id
+
+        raise MutterConfigError('Monitor mode with selected resolution not found')
+
+    def to_dbus(self):
+        return (
+            self.x,
+            self.y,
+            self.scale,
+            self.transform,
+            self.primary,
+            self.monitors,
+        )
+
+
+class MutterDisplay(object):
+    """Class wrapping Mutter's display configuration API."""
+
+    def __init__(self):
+        self._proxy = MUTTER_DISPLAY_CONFIG.get_proxy()
+
+    def on_service_ready(self, callback):
+        observer = DBusObserver(SessionBus, 'org.gnome.Kiosk')
+        observer.service_available.connect(callback)
+        observer.connect_once_available()
+
+    def set_resolution(self, res_str):
+        """Changes the screen resolution.
+
+        :param res_str: Screen resolution configuration with format "800x600".
+        :raises MutterConfigError on failure.
+        """
+        if not self._proxy.ApplyMonitorsConfigAllowed:
+            raise MutterConfigError('Monitor configuration is not allowed')
+
+        (width, height) = self._parse_resolution_str(res_str)
+        (serial, monitor_props, logical_monitor_props, _) = self._proxy.GetCurrentState()
+
+        persistent_config = 2
```

Comment:
I think it would be good to document what this value does & where it comes from.

Reply:
It is documented in the D-Bus method ApplyMonitorsConfig. However, since dasbus doesn't require the XML definition (with the docs) better document it here as well :+1: